autosklearn/data/feature_validator.py

```diff
@@ -0,0 +1,399 @@
+        # That is, this attribute contains whether autosklearn
+        # should treat a column as categorical or numerical
+        # During fit, if the user provided feat_types, the user
+        # constrain is honored. If not, this attribute is used.
+        self.feature_types = None  # type: typing.Optional[typing.List[str]]
+
+        self.data_type = None  # type: typing.Optional[type]
+
+        self.encoder = None  # type: typing.Optional[BaseEstimator]
+        self.enc_columns = []  # type: typing.List[str]
+
+        self._is_fitted = False
+
+    def fit(
+        self,
+        X_train: typing.Union[np.ndarray, pd.DataFrame],
+        X_test: typing.Optional[typing.Union[np.ndarray, pd.DataFrame]] = None,
+        feat_type: typing.Optional[typing.List[str]] = None,
+    ) -> BaseEstimator:
+        """
+        Validates and fit a categorical encoder (if needed) to the features.
+        The supported data types are List, numpy arrays and pandas DataFrames.
+        CSR sparse data types are also supported
+
+        Parameters
+        ----------
+            X_train: typing.Union[np.ndarray, pd.DataFrame]
+                A set of features that are going to be validated (type and dimensionality
+                checks) and a encoder fitted in the case the data needs encoding
+            X_test: typing.Union[np.ndarray, pd.DataFrame]
+                A hold out set of data used for checking
+            feat_type: typing.Optional[typing.List[str]]
+                In case the data is not a pandas DataFrame, this list indicates
+                which columns should be treated as categorical
+        """
+        # Check that the data is valid
+        self.check_data(X_train, X_test, feat_type)
+
+        # Fit on the training data
+        self._fit(X_train)
+
+        self._is_fitted = True
+
+        return self
+
+    def _fit(
+        self,
+        X: typing.Union[np.ndarray, pd.DataFrame],
+    ) -> BaseEstimator:
+        """
+        In case input data is a pandas DataFrame, this utility encodes the user provided
+        features (from categorical for example) to a numerical value that further stages
+        will be able to use
+
+        Parameters
+        ----------
+            X: typing.Union[np.ndarray, pd.DataFrame]
+                A set of features that are going to be validated (type and dimensionality
+                checks) and a encoder fitted in the case the data needs encoding
+        """
+        if hasattr(X, "iloc") and not scipy.sparse.issparse(X):
+            # Treat a column with all instances a NaN as numerical
+            if np.any(pd.isnull(X)):
+                for column in X.columns:
+                    if X[column].isna().all():
+                        X[column] = pd.to_numeric(X[column])
+
+            self.enc_columns, self.feature_types = self._get_columns_to_encode(X)
+
+            if len(self.enc_columns) > 0:
+
+                self.encoder = make_column_transformer(
```

Comment:
I'm really wondering whether we should also encode numpy arrays once the OrdinalEncoder supports missing values. I'm also wondering what we do with lists of lists, don't they need to be encoded as well?